openshift-knative-operator/pkg/eventing/extension.go

```diff
@@ -0,0 +1,41 @@
+package eventing
+
+import (
+	"context"
+	"os"
+
+	mf "github.com/manifestival/manifestival"
+	"github.com/openshift-knative/serverless-operator/openshift-knative-operator/pkg/common"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+	operator "knative.dev/operator/pkg/reconciler/common"
+)
+
+// NewExtension creates a new extension for a Knative Eventing controller.
+func NewExtension(ctx context.Context) operator.Extension {
+	return &extension{}
+}
+
+type extension struct{}
+
+func (e *extension) Transformers(v1alpha1.KComponent) []mf.Transformer {
+	return nil
+}
+
+func (e *extension) Reconcile(ctx context.Context, comp v1alpha1.KComponent) error {
+	ke := comp.(*v1alpha1.KnativeEventing)
+
+	// Override images.
+	images := common.ImageMapFromEnvironment(os.Environ())
+	ke.Spec.Registry.Override = images
+	ke.Spec.Registry.Default = images["default"]
```

Comment:
This might violate "The Principle of Least Surprise". If we want to make the deliberate decision to ignore any values the user might put in their CR, for whatever reason, I don't think we should do it silently.

My personal preference would be to let their spec fields, if present, win, but at the very least I think we should toss an error if they've set any field[s] we won't honor.

Otherwise, SURPRISE!

Reply:
I agree in principle, though I was trying to preserve behavior that we are exhibiting currently. We're currently hard-overriding the `Override` section and are "override-if-present" the default and queue-proxy section.

I thought it'd be clearer if we just override the entire registry as those small exceptions didn't seem worth it at all.

I'd rather revisit our behavior separately once the new logic is engaged to keep the amount of moving parts small. This allows us to replace the components while not expecting any significant behavioral changes if that makes sense?

Reply:
For the registry specifically we don't really have a mechanism not to override, since on upgrade this will contain outdated images and **has** to be overridden.

Reply:
yeah, good point :-)

Reply:
Added TODOs and a JIRA to capture the very valid sentiment here. I believe we should (and can!) revisit after we're through with the transition and can actually drop the webhook defaulting.

Reply:
> I was trying to preserve behavior that we are exhibiting currently

+1 on this approach! Let's do not try to change the architecture and improve units at the same time!